src/sagemaker_xgboost_container/checkpointing.py

```diff
@@ -0,0 +1,317 @@
+                    os.remove(path)
+                except Exception:
+                    logger.debug("Failed to delete %s", path)
+                finally:
+                    self.delete_queue.task_done()
+            self.delete_queue.task_done()
+
+        def _delete_twice():
+            _delete_once(skip_locked_files=True)
+            # Here, we've reached the end of training because we place sentinel in the
+            # queue at the end of training. We put another sentinel, go through everything
+            # in the queue once again, and try to remove it anyway where there is a lock on
+            # the file or not, because the training is done. On sagemaker, this should send
+            # a delete signal to the agent so that the upload can be canceled and removed
+            # from S3.
+            self.delete_queue.put(self.SENTINEL)
+            _delete_once(skip_locked_files=False)
+
+        self.thread = threading.Thread(target=_delete_twice, daemon=True)
+        self.thread.start()
+
+    def stop(self):
+        """Stops the background thread"""
+        # placing a sentinel in the queue signals that training has ended.
+        self.delete_queue.put(self.SENTINEL)
+        # training has ended, so join the background back with the main thread.
+        self.thread.join()
+
+    def _save_checkpoint(self, model, iteration):
+        """Saves checkpoint to a file path formatted with iteration number"""
+        with tempfile.NamedTemporaryFile(
+                dir=self.checkpoint_dir, suffix=TEMP_FILE_SUFFIX, delete=False) as tf:
+            model.save_model(tf.name)
+
+        save_file_path = self.format_path(iteration)
+        os.rename(tf.name, save_file_path)
+
+    def callback(self, env):
+        # env.rank: rabit rank of the node/process. master node has rank 0.
+        # env.iteration: current boosting round
+        # env.begin_iteration: round # when training started. this is always 0.
+        # env.end_iteration: round # when training will end. this is always num_round + 1.
+        # env.model: model object
+        if env.rank != 0:
+            logger.debug("Not master (rank = %d). Exiting checkpoint callback.", env.rank)
+            return
+
+        current_iteration = self.start_iteration + env.iteration
+        self._save_checkpoint(env.model, current_iteration)
+
+        # At the end of each iteration, we save each checkpoint to different files
+        # parametrized by the iteration number, e.g., checkpoint.1, checkpoint.2,
+        # etc. However, this could result in a large number of files to save locally
+        # (and on S3 if used in SageMaker). To save disk space and reduce the amount
+        # of data required to be downloaded on resumption, we retain only the N
+        # (default 5, spcified by 'max_to_keep') most recent checkpoints.
+        file_to_delete = self.format_path(current_iteration - self.max_to_keep)
+        file_to_delete_exists = os.path.isfile(file_to_delete)
```

Comment:
Instead of checking this here, I would just check during the actual delete. For example, just because the file exists now doesn't necessarily mean it will exist at deletion time, so this check isn't very protective.

Reply:
Thanks for this. I think this is a good suggestion.